### PR TITLE
Hutchinson Trace Estimator with HutchWorkspace and Tests

### PR DESCRIPTION
**Pull request by akij17**
I have added HutchWorkspace which supports user-defined random functions. 
For example: `f(n) = cu(rand(-1.0:2.0:1.0, n))`
I have also added Unit Tests for hutchinson.jl. The estimator does not work with high condition number matrices resulting in failure of one of the test case.

### CODE REVIEW

File: src/diagapp.jl [hunk 1]
```diff
@@ -1,25 +1,37 @@
 # Diagonal Approximation for trace of inverse of matrix
+# Values from Extrapolation Methods for Estimating Trace of the Matrix Inverse by P. Fika
 
-export diagapp
+export diagapp, diagapp!
 
 using LinearAlgebra
 
+struct diagappspace
+    A::AbstractArray{<:Any, 2}
```
Comment (mohdibntarek) [inline src/diagapp.jl]: Use type parameters here.

File: src/diagapp.jl [hunk 2]
```diff
@@ -1,25 +1,37 @@
 # Diagonal Approximation for trace of inverse of matrix
+# Values from Extrapolation Methods for Estimating Trace of the Matrix Inverse by P. Fika
 
-export diagapp
+export diagapp, diagapp!
 
 using LinearAlgebra
 
+struct diagappspace
```
Comment (mohdibntarek) [inline src/diagapp.jl]: Convention in struct names is upper case first letter.

File: src/hutchinson.jl [hunk 3]
```diff
@@ -1,59 +1,135 @@
...
 # Values from Extrapolation Methods for Estimating Trace of the Matrix Inverse by P. Fika
+# (This method works for SPD with low condition number)
 
-export hutch
+export hutch, hutch!, HutchWorkspace
 
 using LinearAlgebra
 
+struct HutchWorkspace
+    A::AbstractArray{<:Any, 2}
```
Comment (mohdibntarek) [inline src/hutchinson.jl]: Use type parameters.

File: src/hutchinson.jl [hunk 4]
```diff
@@ -1,59 +1,135 @@
...
+    randfunc::Function
```
Comment (mohdibntarek) [inline src/hutchinson.jl]: Use type parameter.

File: src/diagapp.jl [hunk 5]
```diff
@@ -1,25 +1,37 @@
...
+    A::AbstractArray{<:Any, 2}
+    x::AbstractArray{<:Any, 1}
+    y::AbstractArray{<:Any, 1}
+end
+
+function diagappspace(A::AbstractArray{<:Any, 2})
+    x = rand(-1.0:2.0:1.0, size(A)[1])
```
Comment (mohdibntarek) [inline src/diagapp.jl]: Style: `size(A)[1]` -> `size(A, 1)`

Comment (mohdibntarek) [inline src/diagapp.jl]: Make the `eltype(x)` the same as `eltype(A)`

File: src/diagapp.jl [hunk 6]
```diff
@@ -1,25 +1,37 @@
...
+    y = similar(x, Float64)
```
Comment (mohdibntarek) [inline src/diagapp.jl]: No need for `Float64` here, assuming `A` has some floating point `eltype`. If `eltype(A) <: Integer`, fall back on `Float64`.

Comment (mohdibntarek) [reply src/diagapp.jl]: This is to support `Float32` and higher precision number types.

File: src/diagapp.jl (hunk 2 above)
Comment (akij17) [reply src/diagapp.jl]: Okay, in my head I actually left it like that since I was gonna keep it for internal use only.

File: src/diagapp.jl (hunk 1 above)
Comment (akij17) [reply src/diagapp.jl]: Sorry, can you give a quick example? (Will correct it for the rest of the errors as well)

File: src/diagapp.jl [hunk 7]
```diff
@@ -1,25 +1,37 @@
...
+    c3 = dot(w.y, w.x)    
 
     p = (c0 * c2)/(c1^2)
     
     v0 = log10((c1^2)/(c0 * c2)) / log10((c1 * c3)/(c2^2))
 end
+
 # Calculating the Approximation for ith value of inverse diagonal
 function dfun(A, i, v)
     isum = 0.0
```
Comment (mohdibntarek) [inline src/diagapp.jl]: `isum = zero(eltype(A))`

File: src/diagapp.jl [hunk 8]
```diff
@@ -33,26 +45,27 @@ function dfun(A, i, v)
...
+    return diagapp!(A) 
+end
+
+"""
+    diagapp!(A::AbstractArray)
+
+Diagonal Approximation algorithm for inverse of matrix (SPD low condition number).
+"""
+function diagapp!(A)
     tr = 0.0
```
Comment (mohdibntarek) [inline src/diagapp.jl]: `tr = zero(eltype(A))`

File: src/hutchinson.jl [hunk 9]
```diff
@@ -1,59 +1,135 @@
...
+"""
+    HutchWorkspace(A; N = 30, skipverify = false)
+
+# Arguments
+ - `A` : Symmetric Hermitian Matrix with low condition number
+ - `N` : Number of iterations (Default: 30)
+ - `skipverify` : If false, it will check isposdef(A) (Default: false)
+"""
+function HutchWorkspace(A; N = 30, skipverify = false)
+    randfunc(n) = rand(-1.0:2.0:1.0, n)
```
Comment (mohdibntarek) [inline src/hutchinson.jl]: Make this return a vector with the same `eltype` as `A`.

File: src/hutchinson.jl [hunk 10]
```diff
@@ -1,59 +1,135 @@
...
+    x = randfunc(size(A)[1])
+    y = similar(x)
+    return HutchWorkspace(A, randfunc, x, y, N, skipverify)
+end
+
+"""
+    HutchWorkspace(A, randfunc::Function; N = 30, skipverify = false)
+
+# Arguments
+ - `A` : Symmetric Hermitian Matrix with low condition number
```
Comment (mohdibntarek) [inline src/hutchinson.jl]: Symmetric is redundant, you probably mean positive definite.

File: src/hutchinson.jl [hunk 11]
```diff
@@ -1,59 +1,135 @@
...
+"""
+function HutchWorkspace(A; N = 30, skipverify = false)
+    randfunc(n) = rand(-1.0:2.0:1.0, n)
+    x = randfunc(size(A)[1])
+    y = similar(x)
+    return HutchWorkspace(A, randfunc, x, y, N, skipverify)
+end
+
+"""
+    HutchWorkspace(A, randfunc::Function; N = 30, skipverify = false)
```
Comment (mohdibntarek) [inline src/hutchinson.jl]: Assume `Base.rand` is passed.

File: src/hutchinson.jl [hunk 12]
```diff
@@ -1,59 +1,135 @@
...
+end
+
 # Calculating the moments and extrapolating c-1 
-function ev(A)
-    # Create a vector with +-1 values with 0.5 probability 
-    x = rand(-1:2:1, size(A)[1])
-    
+function ev(w::HutchWorkspace)
+    # Create new random values for x
+    copyto!(w.x, w.randfunc(size(w.A)[1]))
```
Comment (mohdibntarek) [inline src/hutchinson.jl]: This line will allocate for the output of `w.randfunc`. If `Base.rand` is passed, we can do:
```julia
T = eltype(w.x)
o = one(T)
t = 2*o
w.x .= rand.(Ref(-o:t:o))
```
Try to understand the last line. `Ref` "protects" the range when broadcasting. Note that other `rand` functions for special array types support broadcasting too, so this should work for `CuArrays.rand` too.

File: src/hutchinson.jl [hunk 13]
```diff
@@ -1,59 +1,135 @@
...
+    hutch!(w::HutchWorkspace; aitken = false)
+    hutch!(A::AbstractArray; N = 30, skipverify = false, aitken = false)
 
-=#
+Take a HutchWorkspace object as input, apply hutchinson estimation algorithm and solve for trace 
+of inverse of the matrix. (in-place version of hutch)
+"""
+function hutch!(w::HutchWorkspace; aitken = false)
+    if w.skipverify == true || isposdef(w.A)
+        tr = 0.0
```
Comment (mohdibntarek) [inline src/hutchinson.jl]: `0.0` is `Float64`, use `zero(T)` where `T` is `eltype(A)`.

File: test/diagapp.jl [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            println("Executing Test 01: Dense SPD Small Size")
+            A = rand(610, 610)
+            for i in 1:610
+                for j in 1:610
+                    A[i, j] = exp(-2 * abs(i - j))
+                end
+            end
+            obv = diagapp!(A)
+            acv = tr(inv(A))
+            @test percent_error(obv, acv)
```
Comment (mohdibntarek) [inline test/diagapp.jl]: Use `isapprox`.

File: src/diagapp.jl [hunk 15]
```diff
@@ -33,26 +45,27 @@ function dfun(A, i, v)
...
 function diagapp(A)
+    return diagapp!(A) 
+end
+
+"""
+    diagapp!(A::AbstractArray)
+
+Diagonal Approximation algorithm for inverse of matrix (SPD low condition number).
+"""
+function diagapp!(A)
```
Comment (mohdibntarek) [inline src/diagapp.jl]: Why the `!`? In Julia, this usually means that `A` is modified somehow, which it is not.

Comment (akij17) [reply src/diagapp.jl]: Yes, I had some confusion for this one. Will fix that asap.

**Pull request merged**